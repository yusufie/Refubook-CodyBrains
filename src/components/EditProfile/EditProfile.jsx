import { useSelector } from 'react-redux';
import { useTranslation } from 'react-i18next';
import { selectUser } from '../../features/userSlice';
import useEditProfile from '../../hooks/useUpdateProfile';
import useAvatarUpload from '../../hooks/useAvatarUpload';
import Loading from '../Icons/Loading';
import camera from '../../assets/camera.png';

const EditProfile = ({ setIsEditing }) => {
  const { t } = useTranslation();

  const user = useSelector(selectUser);

  const { file, handleFileChange, photoURL, loading, percentage } =
    useAvatarUpload();

  const { register, handleSubmit, onSubmit, errors, reset } = useEditProfile();

  const handleCancel = () => {
    setIsEditing(false);
    reset();
  };

  const roundedPercentage = Math.round(percentage);

  return (
    <div className="flex flex-col items-center justify-center gap-4">
      <form
        onSubmit={handleSubmit((data) =>
          onSubmit(data, photoURL, setIsEditing)
        )}
        className="flex flex-col items-center justify-center gap-4"
      >
        <div className="flex flex-col items-center justify-center relative">
          {loading && file ? (
            <Loading />
          ) : (
            <img
              src={photoURL || user.photoURL}
              alt="avatar"
              className="w-40 h-40 rounded-full"
            />
          )}

          <label
            htmlFor="photoURL"
            className="absolute bottom-0 right-0 bg-refubookBlue p-2 rounded-full cursor-pointer"
          >
            <img src={camera} alt="" className="w-6 h-6" />
            <input
              type="file"
              id="photoURL"
              name="photoURL"
              className="hidden"
              {...register('photoURL', {
                value: photoURL,
              })}
              onChange={handleFileChange}
            />
          </label>

          {errors.photoURL && (
            <span className="text-red-500 text-sm">
              {errors.photoURL?.message}
            </span>
          )}
        </div>
        <div className="flex flex-col items-center justify-center gap-2 w-full">
          {percentage > 0 && (
            <div className="w-full flex mt-4 h-2 bg-gray-300 rounded-full max-w-xs">
              <div
                className="h-full bg-refubookBlue rounded-full"
                style={{ width: `${percentage}%` }}
              />
              <p className="text-center">{roundedPercentage}%</p>
            </div>
          )}
        </div>
        <div className="flex flex-col md:flex-row items-center justify-evenly gap-2 md:gap-10 w-full">
          <label
            htmlFor="name"
            className="text-refubookBlue w-full font-medium text-lg"
          >
            {t('Name')}:
            <input
              type="text"
              id="name"
              className="w-full border border-refubookGray p-2 rounded-md text-refubookBlack"
              {...register('name')}
            />
          </label>
          {errors.name && (
            <span className="text-red-500 text-sm">{errors.name.message}</span>
          )}
          <label
            htmlFor="surname"
            className="text-refubookBlue w-full font-medium text-lg"
          >
            {t('Surname')}:
            <input
              type="text"
              id="surname"
              className="w-full border border-refubookGray p-2 rounded-md text-refubookBlack"
              {...register('surname')}
            />
          </label>
          {errors.surname && (
            <span className="text-red-500 text-sm">
              {errors.surname.message}
            </span>
          )}
        </div>
        <div className="flex  items-center justify-center gap-2 w-full">
          <label
            htmlFor="biography"
            className="text-refubookBlue w-full font-medium text-lg"
          >
            {t('Biography')}:
            <textarea
              id="biography"
              name="biography"
              className="w-full border border-refubookGray p-1 rounded-md text-refubookBlack"
              {...register('biography')}
            />
          </label>
          {errors.biography && (
            <span className="text-red-500 text-sm">
              {errors.biography.message}
            </span>
          )}
        </div>
        <div className="flex  items-center justify-center gap-2 w-full">
          <label
            htmlFor="location"
            className="text-refubookBlue w-full font-medium text-lg"
          >
            {t('Location')}:
            <input
              type="text"
              id="location"
              name="location"
              className="w-full border border-refubookGray p-2 rounded-md text-refubookBlack"
              {...register('location')}
            />
          </label>
          {errors.location && (
            <span className="text-red-500 text-sm">
              {errors.location.message}
            </span>
          )}
        </div>
        <div className="flex  items-center justify-evenly gap-2 w-full max-w-xs">
          <button
            type="submit"
            className=" group relative flex w-28 justify-center rounded-full border border-transparent bg-refubookBlue py-2 px-4  text-sm font-medium mt-8 text-refubookWhite hover:bg-refubookLightBlue focus:outline-none focus:ring-2 focus:ring-[#331416] focus:ring-offset-2"
          >
            {t('Save')}
          </button>
          <button
            type="button"
            className="group relative flex w-28 justify-center rounded-full border border-transparent bg-refubookWhite py-2 px-4 text-sm font-medium mt-8 text-refubookBlack hover:bg-refubookLightBlue focus:outline-none focus:ring-2 focus:ring-[#331416] focus:ring-offset-2"
            onClick={handleCancel}
          >
            {t('Cancel')}
          </button>
        </div>
      </form>
    </div>
  );
};

export default EditProfile;
