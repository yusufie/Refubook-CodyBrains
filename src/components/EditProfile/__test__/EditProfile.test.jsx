import renderer from 'react-test-renderer';
import EditProfile from '../EditProfile';

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: () => jest.fn(),
  useDispatch: () => jest.fn(),
}));

jest.mock('../../../hooks/useUpdateProfile', () => ({
  __esModule: true,
  default: () => ({
    register: jest.fn(),
    handleSubmit: jest.fn(),
    onSubmit: jest.fn(),
    errors: jest.fn(),
    reset: jest.fn(),
  }),
}));

jest.mock('../../../hooks/useAvatarUpload', () => ({
  __esModule: true,
  default: () => ({
    file: jest.fn(),
    handleFileChange: jest.fn(),
    photoURL: jest.fn(),
    loading: jest.fn(),
    percentage: jest.fn(),
  }),
}));

// eslint-disable-next-line no-console
const setIsEditing = jest.fn(() => console.log('setIsEditing called'));

describe('EditProfile component', () => {
  it('should match snapshot', () => {
    const tree = renderer
      .create(<EditProfile setIsEditing={setIsEditing} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
