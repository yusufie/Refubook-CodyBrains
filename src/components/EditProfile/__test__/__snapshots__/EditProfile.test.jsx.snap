// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditProfile component should match snapshot 1`] = `
<div
  className="flex flex-col items-center justify-center gap-4"
>
  <form
    className="flex flex-col items-center justify-center gap-4"
  >
    <div
      className="flex flex-col items-center justify-center relative"
    >
      <svg
        className="animate-spin w-32 h-32 rounded-full"
        version="1.1"
        viewBox="0 0 100 100"
        x="0"
        xmlSpace="preserve"
        xmlns="http://www.w3.org/2000/svg"
        y="0"
      >
        <path
          d="M73 50c0-12.7-10.3-23-23-23S27 37.3 27 50m3.9 0c0-10.5 8.5-19.1 19.1-19.1S69.1 39.5 69.1 50"
          fill="#4699C2"
        >
          <animateTransform
            attributeName="transform"
            attributeType="XML"
            dur="1s"
            from="0 50 50"
            repeatCount="indefinite"
            to="360 50 50"
            type="rotate"
          />
        </path>
      </svg>
      <label
        className="absolute bottom-0 right-0 bg-refubookBlue p-2 rounded-full cursor-pointer"
        htmlFor="photoURL"
      >
        <img
          alt=""
          className="w-6 h-6"
          src="camera.png"
        />
        <input
          className="hidden"
          id="photoURL"
          name="photoURL"
          onChange={[MockFunction]}
          type="file"
        />
      </label>
    </div>
    <div
      className="flex flex-col items-center justify-center gap-2 w-full"
    />
    <div
      className="flex flex-col md:flex-row items-center justify-evenly gap-2 md:gap-10 w-full"
    >
      <label
        className="text-refubookBlue w-full font-medium text-lg"
        htmlFor="name"
      >
        Name
        :
        <input
          className="w-full border border-refubookGray p-2 rounded-md text-refubookBlack"
          id="name"
          type="text"
        />
      </label>
      <span
        className="text-red-500 text-sm"
      />
      <label
        className="text-refubookBlue w-full font-medium text-lg"
        htmlFor="surname"
      >
        Surname
        :
        <input
          className="w-full border border-refubookGray p-2 rounded-md text-refubookBlack"
          id="surname"
          type="text"
        />
      </label>
    </div>
    <div
      className="flex  items-center justify-center gap-2 w-full"
    >
      <label
        className="text-refubookBlue w-full font-medium text-lg"
        htmlFor="biography"
      >
        Biography
        :
        <textarea
          className="w-full border border-refubookGray p-1 rounded-md text-refubookBlack"
          id="biography"
          name="biography"
        />
      </label>
    </div>
    <div
      className="flex  items-center justify-center gap-2 w-full"
    >
      <label
        className="text-refubookBlue w-full font-medium text-lg"
        htmlFor="location"
      >
        Location
        :
        <input
          className="w-full border border-refubookGray p-2 rounded-md text-refubookBlack"
          id="location"
          name="location"
          type="text"
        />
      </label>
    </div>
    <div
      className="flex  items-center justify-evenly gap-2 w-full max-w-xs"
    >
      <button
        className=" group relative flex w-28 justify-center rounded-full border border-transparent bg-refubookBlue py-2 px-4  text-sm font-medium mt-8 text-refubookWhite hover:bg-refubookLightBlue focus:outline-none focus:ring-2 focus:ring-[#331416] focus:ring-offset-2"
        type="submit"
      >
        Save
      </button>
      <button
        className="group relative flex w-28 justify-center rounded-full border border-transparent bg-refubookWhite py-2 px-4 text-sm font-medium mt-8 text-refubookBlack hover:bg-refubookLightBlue focus:outline-none focus:ring-2 focus:ring-[#331416] focus:ring-offset-2"
        onClick={[Function]}
        type="button"
      >
        Cancel
      </button>
    </div>
  </form>
</div>
`;
